### DISCUSSION TITLE
Ability to reinitialize a formset and input widgets after page has loaded?

### DISCUSSION THREAD

**Discussion by SamuelJennings** [jrief/django-formset#100] [Ideas]
Hi jrief,

I'm trying to add some some interactivity to my app using [htmx](https://htmx.org) to inject forms into an existing web page. I have a formset that works fine when I use it normally, however, if I inject that same form using htmx, some of the widgets (e.g. selectize, rich text, etc.) do not render properly on the page.

Not raising an issue because this is probably outside the expected use case of django-formset. But perhaps you have some ideas? Is there a way to re-initialize a formset after the page has already loaded?

**Normal use (works fine):**

![image](https://github.com/jrief/django-formset/assets/47405990/afb9840e-ae42-445e-8841-3b960b9b10c5)
  
**HTMX injection after page load (not so good):**

![image](https://github.com/jrief/django-formset/assets/47405990/07badb61-b8d4-48b1-ae7c-c66285d1f150)

**Comment by jrief** [comment]
Hi @SSJenny90 ,
great to hear that you attempted to get django-formset up and running with HTMX. This actually is something on my TODO list because I believe that those two technologies fit well together. However, I do not have a lot of experience with HTMX. Could you please create a fork and add your code into a demo below `testapp/`. I then can check and look for a solution.

**Comment by SamuelJennings** [comment]
(reposting as a top level comment)

Totally agree that both technologies fit well together. Here's a demo app as requested:

https://github.com/SSJenny90/django-formset

It adds an additional HTMXFormView that shows up as "htmx" under the index view of each framework. I configured it with the OpinionForm because that was the first I found that utilized the Selectize widgets in order to demonstrate the issue. The new view should happily work with any other form as required.

If you need anything else or require more explanation, let me know.

**Reply by jrief** [reply]
It will take a little bit to implement this. Hope you're not in a hurry.

**Reply by SamuelJennings** [reply]
No great rush on my part. Take your time.

**Comment by mpclarkson** [comment]
This is a truly exceptional library @jrief... thought all my dreams had come truebut I too use HTMX extensively and it's not at all straightforward. Well done though. Great work.

**Reply by mpclarkson** [reply]
What we need is something like `DjangoFormset.init` to reinitialise the JS on an htmx:swap event. Then my life would be complete.

**Comment by jrief** [comment]
What should such a `DjangoFormset.init(…)`-function do?

I actually am more focused on using the PEG-parser to create my own triggers in analogy to the HTMX triggers. Maybe you could describe some use cases, how your project works.

I currently have a long backlog, so please be patient.

**Reply by mpclarkson** [reply]
@jrief basically I need to be able to do something like this to reinitialise django-formset when partial templates are loaded via HTMX. 

```
htmx.onLoad(function(content) {
  enableUIElements();
});
```
Currently, if you load a django-formset in a modal presented using HTMX it doesn't work as the entire dom is not updated.

**Comment by SamuelJennings** [comment]
I don't normally venture far into the JS/TS world but I took a look at django-formset.ts and django-formset.monlith.ts, and noticed that all of the init logic happens directly inside a 'DOMContentLoaded' listener. Would it be possible to pull this code out into a function and then expose the function to the global window object?

Something like this: 

```javascript
// django-formset.ts

function initializeFormset() {
   // all initiliazation logic here
}

window.addEventListener('DOMContentLoaded', (event) => {
    intializeFormset();
}

// Attach the function to the global `window` object
window.initializeFormset = initializeFormset;
```

In this case, all someone needs to do in their page is recall the initializeFormset method when it suits them (e.g. inside an HTMX event listener). I'm sure there's more to it than this but maybe it gets us moving towards a solution.

**Reply by mpclarkson** [reply]
Yes, this should work 💯
